### DISCUSSION TITLE
ImageGlass Moon 9.0 beta 1

### DISCUSSION THREAD

**Discussion by d2phap** [d2phap/ImageGlass#1338] [General]
<img width="640" alt="ImageGlass 9.0 beta 1" src="https://github.com/ImageGlass/config/blob/main/screenshots/v9.0-beta-1/9.0b1_1.jpg?raw=true">

[![Downloads](https://img.shields.io/github/downloads/d2phap/imageglass/9.0.1.514/total?color=%232A7C91&label=downloads&style=for-the-badge)](https://github.com/d2phap/ImageGlass/releases/download/9.0.1.514/ImageGlass_Moon_9.0.1.514_Beta-1_x64.zip)

## ⚠ Breaking changes
This is the first beta release of ImageGlass Kobe version 9, rewritten from scratch. Many features of the version 8 are missing and there are a lot of breaking changes. All features and settings may be changed when the official version 9 is released.

```
💀 If you want to try this bleeding edge release, use it at your own risk!
```

To follow the progress of version 9 beta 2, please follow: https://github.com/d2phap/ImageGlass/projects/12.
You also can support this project by making a donation or purchasing the Microsoft Store version (version 8):

<a href="https://www.microsoft.com/store/productId/9N33VZK3C7TH">
  <img height="56" src="https://upload.wikimedia.org/wikipedia/commons/f/f7/Get_it_from_Microsoft_Badge.svg" />
</a>

## 🎐 Minimum requirements
- [.NET 6.0 Runtime](https://dotnet.microsoft.com/en-us/download/dotnet/6.0)
- [Windows 7 SP1 ESU](https://docs.microsoft.com/en-us/dotnet/core/install/windows?tabs=net60) 64 bit (32-bit version will be added when the official version 9 is release)

## 🌩 Highlights
- Replaces all XML files by JSON files.
- Supports hardware acceleration for non-alpha, non-animating image formats (#272)
- Adds Image Focus mode for better panning and zooming the viewing image
- Supports hotkey customization for menus, and Image Focus mode (#608)
- Supports more flexible customization for toolbar buttons and positions
- Supports smooth scrolling and displaying tooltip for thumbnails
- Adds ability to choose which image information tags to display on title bar (#971)
- Adds more options in theme pack
- Supports FITS, XV, VIFF, MJPEG and animating WEBP format (#508)
- The server for Check for update feature is changed to https://github.com/ImageGlass/config
- Changes registry keys for Set as default photo viewer from `...\PhapSoftware\ImageGlass\...` to `...\ImageGlass<code>\...`, while `code` is either `Kobe`, or `Moon`. This is ImageGlass Moon release, so the key is `...\ImageGlassMoon\...`.

## Limitations
- Changing settings in runtime is not supported yet. If you want to change settings, you can do it using `igconfig.json` before the app runs.
- Supporting for animating WEBP format is not native yet, ImageGlass converts WEBP to GIF to play the animation.
- Hardware acceleration is not supported for those images that contain alpha pixels (transparent) and smaller than 16,380 pixel width and height.

## 🏀 Supports features from v8
<details>
<summary>Show/hide details</summary>

* File > Open file..
* File > Open width...
* File > Print...
* File > Refresh / Reload / Reload image list
* Navigation > View next / previous image
* Navigation > View the first / last image
* Zoom > Zoom in / out / view actual size
* Zoom > all 6 zoom modes from version 8: 
  * Auto zoom, 
  * Lock zoom, 
  * Scale to width, 
  * Scale to height, 
  * Scale to fit, 
  * Scale to fill.
* Image > **Image focus mode** (new)
* Image > Loading orders
* Image > Open image location
* Image > View image properties
* Clipboard> Copy / Cut multiple image files
* Clipboard > Copy image pixel data
* Clipboard > Copy image path
* Layout: All menu features from version 8: Toggle toolbar, thumbnail bar, checkerboard background, set window always on top
* Help > **Set as default photo viewer** (new)
* Help > **Unset default photo viewer** (new)
* And other features...

</details>

## 🌌 Full details
<details>
<summary>Show/hide details</summary>

### Image Booster component
- Supports cancelling loading the previous images on fast navigating (#1025, #1107)
- Supports custom codecs. Currently, only contains Magick.NET codec

### Settings, Theme, and Language component
- Change file format to JSON format
- For Settings component:
  - Supports 3 levels of setting: `igconfig.json`, `igconfig.default.json`, `igconfig.admin.json`
  - Supports passing individual settings to `ImageGlass.exe` through command line (#669).
  The setting must start with a hyphen `-`, and setting value comes after equal character `=`. For example:
    ```bash
    # hide the toolbar and set background color to red
    ImageGlass.exe -ShowToolbar=false -BackgroundColor=#ff0000
    ```
  - Image information setting: adds ability to choose which info tags to display on the title bar thought the setting `InfoItems: []` (#971), supports EXIF Rating.

### ViewBox component
- Supports hardware acceleration to improve image interaction performance (#272)
- Navigation arrow buttons are improved with better look and interaction behaviors (#1224)
- Supports showing better text message
- Supports Image focus mode for smooth zooming and panning (#272, #522)
- Able to change zooming/panning speed
- Supports customize keyboard for Image focus mode through the setting `ImageFocusModeHotkeys: {}` (#608). Default values:
  ```json
  "ImageFocusModeHotkeys": {
    "InternalPanningLeftKeys": "Left",
    "InternalPanningRightKeys": "Right",
    "InternalPanningUpKeys": "Up",
    "InternalPanningDownKeys": "Down",
    "InternalZoomInKeys": "-",
    "InternalZoomOutKeys": "+"
  }
  ```

### Toolbar component
- Supports fully customizable toolbar buttons through the setting `ToolbarItems: []`. For example:
  ```js
  "ToolbarItems": [
    {
      "Type": "Button", // "Button" or "Separator"
      "Id": "btn_Thumbnail", // Id must be unique
      "Text": "Thumbnail bar", // Tooltip text for the button
      "DisplayStyle": "Image", // Style of button: "Image" or "ImageAndText"
      "CheckableConfigBinding": "ShowThumbnails", // Bind the setting name for auto-check on click
      "Alignment": "Left", // Alignment group: "Left" or "Right"
      "Image": "ThumbnailBar", // Predefined image key in theme pack
      "OnClick": {
        "Executable": "MnuToggleThumbnails", // Can be the name of menu, name of function in FrmMain.IGMethods, or path of executable file.
        "Argument": "" // argument for function in FrmMain.IGMethods; or command arguments for executable file.
      }
    },
  ]
  ```
- Able to add new buttons with custom actions: Use menu item's action; predefined methods; or a custom command
- Supports 2 position groups: Left/Center group, Right group
- Adds more properties for theming

### Thumbnail bar component
- Adds option to show filename under the thumbnail (#539)
- Supports item tooltip
- Supports dark mode for scrollbar
- Supports persistent disk cache, maximum 100 MB by default (#1292)
- Supports dragging thumbnail to other app (#990)
- Improves scrolling performance by mouse wheel, trackpad.
- Improves toolbar overflow panel (#541)

### Menu component
- Supports custom keyboard shortcut through the setting `MenuHotkeys` (#608). For example:
  ```json
  "MenuHotkeys": {
    "MnuAbout": "Ctrl+Shift+Alt+A",
    "MnuCopyPath": "F1",
    "MnuOpenFile": "O"
  }
  ```
- Removes the ugly top-left lines (#1218)

### Other improvements and fixes
- Upgraded Magick.NET to v11.1.2
- Fixed: when single instance enabled, window always stayed behind when passing a file (#620)
- Uses the new API to save the window position: `WindowPlacement`. This works better with multi-screens

</details>

## 🥒 Checksums:

File | SHA-1
-- | --
ImageGlass_Moon_9.0.1.514_Beta-1_x64.zip | `db24bf1ecdff6c411e4b6dae39c654bb33172ae5`

<hr /><em>This discussion was created from the release <a href='https://github.com/d2phap/ImageGlass/releases/tag/9.0.1.514'>ImageGlass Moon 9.0 beta 1</a>.</em>

**Comment by d2phap** [comment]
Pushed a fix to resolve this issue: 

![image](https://user-images.githubusercontent.com/3154213/168458274-3c424a57-f1dc-4a94-9a39-a3bb8c421c16.png)

**Comment by d2phap** [comment]
Pushed a fix (v9.0.1.518) to resolve this:

![image](https://user-images.githubusercontent.com/3154213/169038707-b0214332-186d-4053-b024-0247e076cdc1.png)

**Comment by PwrSrg** [comment]


**Reply by d2phap** [reply]
Thank you @PwrSrg 
The Beta version is more stable than I thought, I have been using as my main viewer for months.
I have added Fullscreen to the todo list for the Beta 2.

**Comment by User-user00** [comment]
Hello - Just came across this app and wondering where's the delete button in v9?  I never installed v8 and when I click on settings, all I get is an empty json file - with no options at all.  Thanks!

This is my toolbar:
![image](https://user-images.githubusercontent.com/39176113/177345804-8426dca7-4fba-4f83-8aed-57f26f3494fe.png)

**Reply by d2phap** [reply]
v9.0 beta is under development, and rewritten from scratch, so it does not have all features of v8.x

**Reply by User-user00** [reply]
No worries.  Do you have the code handy that I could add to the .json file that would add a "delete" button - or should i just use v8?  Thanks!

**Reply by d2phap** [reply]
Deleting file feature is already implemented for the v9.0 beta 2. If you cannot wait for 2 months, you can use this custom script, or install v8.x.

You can create a `del.bat` file with the content like this:
```bat
del %1
```

Open igconfig.json file of ImageGlass 9.0 beta 1, add this JSON to the `ToolbarItems`:
```js
  "ToolbarItems": [
    // other items here...

    // Add delete button
    {
      "Type": "Button",
      "Id": "btn_Delete",
      "Text": "Delete...",
      "DisplayStyle": "Image",
      "CheckableConfigBinding": "",
      "Alignment": "Left",
      "Image": "Delete",
      "OnClick": {
        "Executable": "path\\to\\del.bat", // full path of del.bat file
        "Argument": "\"<file>\"" // <file> is a macro to pass the current viewing image path to the Executable
      }
    }
  ],
```

After done this, you should be able to see the Delete button on toolbar:

![image](https://user-images.githubusercontent.com/3154213/178181506-c9dab224-c359-4065-bc11-4a3afea69a21.png)

When you click this button, ImageGlass will execute the script in `del.bat` file, which is permanently delete the file from hard disk WITHOUT the confirmation. Also note that, the file is deleted on hard disk but not in the image list of ImageGlass, you will need to manually press Ctrl+Shif+R (Main menu > File > Reload image list) to reload the image list.

**Reply by User-user00** [reply]
OK - got it.  I'll give it a shot and see what happens.  thank you for the detailed response.  Much appreciated!

Edit update:  Works very well.  Thanks!

**Comment by Zeus64** [comment]
Hi, what the status of ImageGlass 9? it's already more than a year I started to hear about ImageGlass 9, is it abandonned ?

**Reply by d2phap** [reply]
No, it's not. I am releasing v9.0 beta 4 soon.

Probably needs half year fpr the official 9.0! There are a lot of works!

**Reply by Zeus64** [reply]
thanks for your job d2phap!
